File: Data Structures & Algorithms/binary-tree-diameter/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int ans = 0;
        function<int(TreeNode*)> go = [&](TreeNode* node) -> int {
            if (!node) return 0;

            int left = go(node->left), right = go(node->right);
            ans = max(ans, left + right);
            return 1 + max(left, right);
        };

        go(root);
        return ans;
        
    }
};

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func isSubtree(root *TreeNode, subRoot *TreeNode) bool {
    /*
        first I have root and subroot pointing to the two roots and present a state so 
        when root.Val == subRoot.Val ---> move the two pointers simultanuosely 
        else move try moving the root to left and right and so on
    */

    if subRoot == nil {
        return true
    }

    if root == nil {
        return false
    }

    if sameTree(root, subRoot) {
        return true
    }

    return isSubtree(root.Left, subRoot) || isSubtree(root.Right, subRoot)
}



func sameTree(root *TreeNode, subRoot *TreeNode) bool {
    if root == nil && subRoot == nil {
        return true
    }

    if root != nil && subRoot != nil && root.Val == subRoot.Val {
        return sameTree(root.Left, subRoot.Left) && sameTree(root.Right, subRoot.Right)
    }

    return false
}
